File: leetcode/bag-of-tokens.py
# 948. Bag of Tokens
# 🟠 Medium
#
# https://leetcode.com/problems/bag-of-tokens/
#
# Tags: Array - Two Pointers - Greedy - Sorting

import timeit
from collections import deque
from typing import List


# Play the smallest tokens to gain score and the bigger ones to gain
# power, that way we can maximize the number of plays.
#
# Time complexity: O(n*log(n)) - Where n is the number of tokens. The
# sorting step has the highest cost, then we can play the tokens in O(n)
# Space complexity: O(n) - The queue has length n.
#
# Runtime: 90 ms, faster than 45.75%
# Memory Usage: 13.9 MB, less than 99.06%
class Solution:
    def bagOfTokensScore(self, tokens: List[int], power: int) -> int:
        # Initialize our score.
        score = 0
        # We want to access both the smallest and biggest element,
        # sort the tokens and use a deque to store them.
        q = deque(sorted(tokens))
        # Keep track of the length of the queue to determine whether we
        # made a move on the last round.
        last_token_count = len(q) + 1
        # Keep going while we have tokens and we made a move on the last
        # round.
        while q and len(q) < last_token_count:
            last_token_count = len(q)
            # If we can score, do it greedily.
            if power >= q[0]:
                # Play face up.
                power -= q.popleft()
                score += 1
            # If we can play a token face down (score > 0) and we are
            # not going to loose points doing so, play the biggest token
            # face down.
            elif score and len(q) > 1:
                # Play the biggest token face down.
                power += q.pop()
                score -= 1
        # Return the result.
        return score


def test():
    executors = [Solution]
    tests = [
        [[], 200, 0],
        [[100], 50, 0],
        [[100, 200], 150, 1],
        [[100, 200, 300, 400], 200, 2],
    ]
    for executor in executors:
        start = timeit.default_timer()
        for _ in range(1):
            for col, t in enumerate(tests):
                sol = executor()
                result = sol.bagOfTokensScore(t[0], t[1])
                exp = t[2]
                assert result == exp, (
                    f"\033[93m» {result} <> {exp}\033[91m for"
                    + f" test {col} using \033[1m{executor.__name__}"
                )
        stop = timeit.default_timer()
        used = str(round(stop - start, 5))
        cols = "{0:20}{1:10}{2:10}"
        res = cols.format(executor.__name__, used, "seconds")
        print(f"\033[92m» {res}\033[0m")


test()

File: leetcode/bag-of-tokens.rs
// 948. Bag of Tokens
// 🟠 Medium
//
// https://leetcode.com/problems/bag-of-tokens/
//
// Tags: Array - Two Pointers - Greedy - Sorting

struct Solution;
impl Solution {
    /// Sort the input, after that greedily iterate over the tokens checking the token at both ends
    /// of the vector, if we have enough remaining power to play the token on the left, do it and
    /// gain one point, otherwise play the token on the right and gain that much power. On each
    /// loop, we check if there are remaining tokens and if our score has gone negative. Return the
    /// maximum amount of points seen.
    ///
    /// Time complexity: O(n*log(n)) - We need to sort the input, after that, O(n)
    /// Space complexity: O(n) - Sorting the input vector and the local copy.
    ///
    /// Runtime 1 ms Beats 100%
    /// Memory 2.08 MB Beats 100%
    pub fn bag_of_tokens_score(mut tokens: Vec<i32>, power: i32) -> i32 {
        let n = tokens.len() as i32;
        tokens.sort_unstable();
        let mut power = power;
        let mut points = 0;
        let mut res = 0;
        let (mut l, mut r) = (0, n - 1);
        while l <= r && points >= 0 {
            if tokens[l as usize] <= power {
                points += 1;
                power -= tokens[l as usize];
                l += 1;
            } else {
                points -= 1;
                power += tokens[r as usize];
                r -= 1;
            }
            res = res.max(points);
        }
        res
    }
}

// Tests.
fn main() {
    let tests = [
        (vec![100], 50, 0),
        (vec![200, 100], 150, 1),
        (vec![71, 55, 82], 54, 0),
        (vec![100, 200, 300, 400], 200, 2),
    ];
    println!("\n\x1b[92m» Running {} tests...\x1b[0m", tests.len());
    let mut success = 0;
    for (i, t) in tests.iter().enumerate() {
        let res = Solution::bag_of_tokens_score(t.0.clone(), t.1);
        if res == t.2 {
            success += 1;
            println!("\x1b[92m✔\x1b[95m Test {} passed!\x1b[0m", i);
        } else {
            println!(
                "\x1b[31mx\x1b[95m Test {} failed expected: {:?} but got {}!!\x1b[0m",
                i, t.2, res
            );
        }
    }
    println!();
    if success == tests.len() {
        println!("\x1b[30;42m✔ All tests passed!\x1b[0m")
    } else if success == 0 {
        println!("\x1b[31mx \x1b[41;37mAll tests failed!\x1b[0m")
    } else {
        println!(
            "\x1b[31mx\x1b[95m {} tests failed!\x1b[0m",
            tests.len() - success
        )
    }
}
